keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,423 @@
+standalone spec and is applied to `pod.spec`. It's only effective when it's not
+nil.
+
+```go
+type PodSpec struct {
+    // TopologySpreadConstraint describes how a group of pods are spread
+    // If specified, scheduler will enforce the constraint
+    // +optional
+    TopologySpreadConstraint *TopologySpreadConstraint
+    ......
+}
+```
+
+#### Option 1
+
+Inside `TopologySpreadConstraint`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
+
+```go
+type TopologySpreadConstraint struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+#### Option 2
+
+Another option is to flatten "required" and "perferred" podAffinityTerms, and
+eliminate embedded "TopologyKey":
+
+```go
+type ScheduleDecision string
+
+const (
+    // do not schedule a pod in all circumstances
+    DoNotSchedule ScheduleDecision = "DoNotSchedule"
+    // schedule a pod despite of any circumstance
+    ScheduleAnyway ScheduleDecision = "ScheduleAnyway"
+)
+
+type TopologySpreadConstraint struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // For example, in a 3-zone cluster, currently pods with the same labelSelector
+    // are spread as 1/1/0:
+    // - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+    // schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
+    // - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey is the key such that we consider each value as a "bucket";
+    // we try to put balanced number of pods into each bucket.
+    TopologyKey string
+    // WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+    // the spreading constraint.
+    // - DoNotSchedule (default) tells the scheduler not to schedule it
+    // - ScheduleAnyway tells the scheduler to still schedule it
+    // Note: it's considered as "Unsatisfiable" only when ActualySkew on all nodes exceeds "MaxSkew".
+    WhenUnsatisfiable ScheduleDecision
```

Comment:
why is this not a bool like required ?

Reply:
+1 to Mayank's question. Do you think you want to support other choices in the future? Do you have any examples in mind?

Reply:
+1 to Mayank's question. Do you think you want to support other choices in the future? Do you have any examples in mind?

Reply:
Right now, nope :)

(I was thinking of a third option to respect NodeSelector/NodeAffinity, but it turns out to be more a reasonable implicit assumption)

@lavalamp are you OK with changing it back to bool: `ScheduleWhenUnsatisfiable bool`.

Reply:
I think things like this are more understandable if the consequences of the action are in the enum value rather than true/false.

E.g., as an outsider, reading "ScheduleWhenUnsatisifiable" I have a lot of questions, like, "does that mean it doesn't schedule unless it's unsatisfiable?"

And I do think that you might need an additional parameter in the future, which is how strong the preference is.

Reply:
> And I do think that you might need an additional parameter in the future, which is how strong the preference is.

That reminds me of a possible scenario called "FallBackToAvailableNodes". In between "ScheduleAnyway" and "DoNotSchedule", maybe it's possible to compute the internal ActualySkew and MinimalSkew among the nodes which passed previous Predicates. This is related with earlier discussion with @bsalamat.

This is just an immature thought though, we definitely won't consider it in initial implementation.

Reply:
I agree with the readability part.
Regarding another parameter to indicate the strength of preference, we already have a weight parameter for all preference (priority) functions.

Reply:
For this particular case, I feel fairly strongly that the strings in the current draft communicate the purpose and implications of the decision much more clearly than the bool.